lib/jekyll/commands/serve.rb

```diff
@@ -274,12 +274,19 @@ def launch_browser(server, opts)
         def boot_or_detach(server, opts)
           if opts["detach"]
             pid = Process.fork do
+              # Detach the process from controlling terminal
+              $stdin.reopen("/dev/null", "r")
+              $stdout.reopen("/dev/null", "w")
+              $stderr.reopen("/dev/null", "w")
```

Comment:
Is it possible to switch this out with the [`Open3`](https://ruby-doc.org/stdlib-2.5.0/libdoc/open3/rdoc/Open3.html) module?

Reply:
If we want to have a simpler or cleaner implementation we have to change the semantics of how this work.

Currently, `--detach` runs these steps:

1. Build the site.
2. Fork into child process and start server.
3. Print the child process PID.
4. Exit main process.

We have several options but they would change the semantics:

The simplest solution is to use `Process.daemon`. It is a single call that does fork and detach stdio then continue rest of the code in forked process. However:
    - If we call `Process.daemon` on main, the problem is that we cannot get and print PID in main process. A typical practice of using this method would be adding a `--pid-file` option to let forked process to write its PID into a file.
    - We can also call `Process.daemon` in the forked process to use it for detach stdio, the problem is that it would fork again and change PID.

Therefore, if we stick with fork based implementation, this is the best we can do unless we agree to change the semantics of how this option works today.

Another option would be replace the fork with a real ruby process that just runs webrick, with two different possible implementations:

1. Serialize webrick options and start webrick with a ruby one-liner using `Process.spawn` (or `open3`).
2. Add an option on Jekyll command line that start webrick without building the site. Spawn a Jekyll process with the same ARGV extended with option that disable the build step, so that it just start the server.

Reply:
A huge thank you for giving me a detailed, well-thought out reply, @ntkme 🙏 🙇

Since I don't intend to change much about this, especially since this is particularly unfamiliar territory for me, the current implementation here will do.